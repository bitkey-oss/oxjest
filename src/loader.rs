use std::path::Path;

use oxc::allocator::Allocator;
use oxc::ast::ast::Program;
use oxc::diagnostics::OxcDiagnostic;
use oxc::parser::{Parser, ParserReturn};
use oxc::semantic::{SemanticBuilder, SemanticBuilderReturn};
use oxc::span::SourceType;
use oxc::transformer::{DecoratorOptions, TransformOptions, Transformer, TransformerReturn};

pub struct Loader;

impl Loader {
    pub fn load_str<'a>(
        &self,
        allocator: &'a Allocator,
        source_text: &'a str,
        source_path: impl AsRef<Path>,
    ) -> Result<Program<'a>, Vec<OxcDiagnostic>> {
        let source_path = source_path.as_ref();
        let source_type = SourceType::from_path(source_path).unwrap();

        let ParserReturn {
            mut program,
            errors,
            ..
        } = Parser::new(allocator, source_text, source_type).parse();
        if !errors.is_empty() {
            return Err(errors);
        }

        let SemanticBuilderReturn { semantic, errors } = SemanticBuilder::new()
            .with_excess_capacity(2.0)
            .build(&program);
        if !errors.is_empty() {
            return Err(errors);
        }

        let (symbols, scopes) = semantic.into_symbol_table_and_scope_tree();
        let options = TransformOptions {
            decorator: DecoratorOptions {
                legacy: true,
                ..Default::default()
            },
            ..Default::default()
        };

        let TransformerReturn { errors, .. } = Transformer::new(allocator, source_path, &options)
            .build_with_symbols_and_scopes(symbols, scopes, &mut program);
        if !errors.is_empty() {
            return Err(errors);
        }

        Ok(program)
    }
}
