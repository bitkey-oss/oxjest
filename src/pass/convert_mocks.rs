use std::iter::once;
use std::sync::atomic::{AtomicUsize, Ordering};

use oxc::allocator::{Box, CloneIn, TakeIn, Vec as ArenaVec};
use oxc::ast::AstBuilder;
use oxc::ast::ast::{
    Argument, BindingPatternKind, Expression, ImportDeclaration, ImportDeclarationSpecifier,
    Program, Span, Statement, VariableDeclaration, VariableDeclarationKind,
};
use oxc::span::GetSpan;
use oxc_traverse::{Traverse, TraverseCtx};

use crate::jest::is_jest_do_mock_call;
use crate::jest::is_jest_mock_call;

fn make_create_mock_factory<'a>(ast: AstBuilder<'a>, id: &'a str) -> Expression<'a> {
    ast.expression_call(
        Span::default(),
        ast.member_expression_static(
            Span::default(),
            ast.expression_identifier(Span::default(), "__oxjest__"),
            ast.identifier_name(Span::default(), "createMockFactory"),
            false,
        )
        .into(),
        Option::<Box<'_, _>>::None,
        ast.vec1(
            ast.expression_await(
                Span::default(),
                ast.expression_import(
                    Span::default(),
                    ast.expression_string_literal(Span::default(), id, None),
                    None,
                    None,
                ),
            )
            .into(),
        ),
        false,
    )
}

fn make_dynamic_import<'a>(
    ast: AstBuilder<'a>,
    decl: &ImportDeclaration<'a>,
    import_name: &'a str,
) -> VariableDeclaration<'a> {
    // __oxjest_import_{}__ = await import("...")
    let await_import = ast.variable_declarator(
        Span::default(),
        VariableDeclarationKind::Const,
        ast.binding_pattern(
            ast.binding_pattern_kind_binding_identifier(Span::default(), import_name),
            Option::<Box<'a, _>>::None,
            false,
        ),
        Some(ast.expression_await(
            Span::default(),
            ast.expression_import(
                Span::default(),
                ast.expression_string_literal(decl.source.span, decl.source.value, decl.source.raw),
                None,
                None,
            ),
        )),
        false,
    );

    // foo = __oxjest_import_{}__.foo, bar = __oxjest_import_{}__.default, ...
    let declarations = decl.specifiers.iter().flatten().map(|specifier| {
        ast.variable_declarator(
            Span::default(),
            VariableDeclarationKind::Const,
            ast.binding_pattern(
                BindingPatternKind::BindingIdentifier(
                    ast.alloc(specifier.local().clone_in(ast.allocator)),
                ),
                Option::<Box<'a, _>>::None,
                false,
            ),
            Some(match specifier {
                ImportDeclarationSpecifier::ImportNamespaceSpecifier(_) => {
                    ast.expression_identifier(Span::default(), import_name)
                }
                _ => ast
                    .member_expression_static(
                        Span::default(),
                        ast.expression_identifier(Span::default(), import_name),
                        match specifier {
                            ImportDeclarationSpecifier::ImportDefaultSpecifier(specifier) => {
                                ast.identifier_name(specifier.span, "default")
                            }
                            ImportDeclarationSpecifier::ImportSpecifier(specifier) => ast
                                .identifier_name(
                                    specifier.imported.span(),
                                    specifier.imported.name(),
                                ),
                            _ => unreachable!(),
                        },
                        false,
                    )
                    .into(),
            }),
            false,
        )
    });

    // const __oxjest_import_{}__ = await import("..."), foo = __oxjest_import_{}__.foo, ...;
    ast.variable_declaration(
        decl.span,
        VariableDeclarationKind::Const,
        ast.vec_from_iter(once(await_import).chain(declarations)),
        false,
    )
}

pub(crate) struct ConvertMocks<'a> {
    mocks: Vec<Expression<'a>>,
}

impl ConvertMocks<'_> {
    pub(crate) fn new() -> Self {
        Self { mocks: Vec::new() }
    }
}

impl<'a, State> Traverse<'a, State> for ConvertMocks<'a> {
    fn exit_program(&mut self, node: &mut Program<'a>, ctx: &mut TraverseCtx<'a, State>) {
        // Insert hoisted mocks at the top of the body
        node.body.splice(
            0..0,
            self.mocks.iter().map(|mock_call| {
                ctx.ast
                    .statement_expression(Span::default(), mock_call.clone_in(ctx.ast.allocator))
            }),
        );

        // Imports don't need to be turned into dynamic imports if there are no mocks
        if !self.mocks.is_empty() {
            let import_id = AtomicUsize::new(0);

            node.body.iter_mut().for_each(|stmt| {
                let Statement::ImportDeclaration(decl) = stmt else {
                    return;
                };

                let import_id = import_id.fetch_add(1, Ordering::Relaxed);
                let import_name = format!("__oxjest_import_{import_id}__");

                *stmt = Statement::VariableDeclaration(ctx.ast.alloc(make_dynamic_import(
                    ctx.ast,
                    decl,
                    ctx.ast.str(&import_name),
                )));
            })
        }
    }

    fn exit_expression(&mut self, node: &mut Expression<'a>, ctx: &mut TraverseCtx<'a, State>) {
        let Expression::CallExpression(call) = node else {
            return;
        };

        let is_jest_mock_call = is_jest_mock_call(call);
        let is_jest_do_mock_call = is_jest_do_mock_call(call);
        if !is_jest_mock_call && !is_jest_do_mock_call {
            return;
        }

        let Expression::StaticMemberExpression(member) = &mut call.callee else {
            // SAFETY: Already checked above
            unreachable!();
        };

        member.property.name = ctx.ast.atom("unstable_mockModule");

        if call.arguments.len() < 2 {
            let Some(Argument::StringLiteral(lit)) = call.arguments.first() else {
                return;
            };

            let id = lit.value.as_str();

            call.arguments
                .push(make_create_mock_factory(ctx.ast, id).into())
        }

        // only jest.mock needs to be hoisted
        if is_jest_mock_call {
            self.mocks.push(node.take_in(ctx.ast.allocator));
        }
    }

    fn exit_statements(
        &mut self,
        node: &mut ArenaVec<'a, Statement<'a>>,
        _ctx: &mut TraverseCtx<'a, State>,
    ) {
        node.retain(|stmt| !matches!(stmt, Statement::ExpressionStatement(stmt) if stmt.expression.is_null()));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::transform;
    use oxc::allocator::Allocator;

    #[test]
    fn test_mock() {
        let source_text = r#"
        import { greet } from "./greeter.js";

        // this mocking needs to be hoisted to the top of this module
        jest.mock("./greeter.js", () => ({
          greet: () => "Hello, world!",
        }));
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, ConvertMocks::new());

        insta::assert_snapshot!(code, @r#"
        jest.unstable_mockModule("./greeter.js", () => ({ greet: () => "Hello, world!" }));
        const __oxjest_import_0__ = await import("./greeter.js"), greet = __oxjest_import_0__.greet;
        "#);
    }

    #[test]
    fn test_mock_auto() {
        let source_text = r#"
        import { greet } from "./greeter.js";

        // this mocking needs to be hoisted to the top of this module
        jest.mock("./greeter.js");
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, ConvertMocks::new());

        insta::assert_snapshot!(code, @r#"
        jest.unstable_mockModule("./greeter.js", __oxjest__.createMockFactory(await import("./greeter.js")));
        const __oxjest_import_0__ = await import("./greeter.js"), greet = __oxjest_import_0__.greet;
        "#);
    }

    #[test]
    fn test_do_mock() {
        let source_text = r#"
        import { greet } from "./greeter.js";

        // this mocking does not need to be hoisted
        jest.doMock("./greeter.js", () => ({
          greet: () => "Hello, world!",
        }));
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, ConvertMocks::new());

        insta::assert_snapshot!(code, @r#"
        import { greet } from "./greeter.js";
        // this mocking does not need to be hoisted
        jest.unstable_mockModule("./greeter.js", () => ({ greet: () => "Hello, world!" }));
        "#);
    }

    #[test]
    fn test_do_mock_auto() {
        let source_text = r#"
        import { greet } from "./greeter.js";

        // this mocking does not need to be hoisted
        jest.doMock("./greeter.js");
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, ConvertMocks::new());

        insta::assert_snapshot!(code, @r#"
        import { greet } from "./greeter.js";
        // this mocking does not need to be hoisted
        jest.unstable_mockModule("./greeter.js", __oxjest__.createMockFactory(await import("./greeter.js")));
        "#);
    }
}
