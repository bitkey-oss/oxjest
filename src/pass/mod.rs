use oxc::allocator::Vec;
use oxc::ast::ast::{Expression, MemberExpression, Program, Statement};
use oxc_traverse::{Traverse, TraverseCtx};

pub(crate) mod hoist_mocks;
pub(crate) mod import_actual;
pub(crate) mod inject_globals;

/// The facade of all transforms combined into one.
/// Be careful with calling order when adding a new transform.
pub(crate) struct Transformer<'a> {
    hoist_mocks: hoist_mocks::HoistMocks<'a>,
    import_actual: import_actual::ImportActual<'a>,
    inject_globals: inject_globals::InjectGlobals,
}

impl Transformer<'_> {
    pub(crate) fn new() -> Self {
        Self {
            hoist_mocks: hoist_mocks::HoistMocks::new(),
            import_actual: import_actual::ImportActual::new(),
            inject_globals: inject_globals::InjectGlobals::new(),
        }
    }
}

impl<'a> Traverse<'a> for Transformer<'a> {
    fn exit_program(&mut self, node: &mut Program<'a>, ctx: &mut TraverseCtx<'a>) {
        self.hoist_mocks.exit_program(node, ctx);
        self.import_actual.exit_program(node, ctx);
        self.inject_globals.exit_program(node, ctx);
    }

    fn exit_expression(&mut self, node: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
        self.hoist_mocks.exit_expression(node, ctx);
        self.import_actual.exit_expression(node, ctx);
    }

    fn enter_member_expression(
        &mut self,
        node: &mut MemberExpression<'a>,
        ctx: &mut TraverseCtx<'a>,
    ) {
        self.inject_globals.enter_member_expression(node, ctx);
    }

    fn exit_statements(&mut self, node: &mut Vec<'a, Statement<'a>>, ctx: &mut TraverseCtx<'a>) {
        self.hoist_mocks.exit_statements(node, ctx);
    }
}
