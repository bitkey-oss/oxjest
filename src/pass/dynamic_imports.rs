use std::iter::once;
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};

use oxc::allocator::{Box, CloneIn};
use oxc::ast::AstBuilder;
use oxc::ast::ast::{
    BindingPatternKind, CallExpression, ImportDeclaration, ImportDeclarationSpecifier, Program,
    Statement, VariableDeclaration, VariableDeclarationKind,
};
use oxc::ast_visit::VisitMut;
use oxc::ast_visit::walk_mut::{walk_call_expression, walk_statement};
use oxc::span::{GetSpan, Span};

use crate::jest::{is_jest_mock_call, is_jest_mock_module_call};
use crate::pass::Pass;

fn make_dynamic_import<'a>(
    ast: AstBuilder<'a>,
    decl: &ImportDeclaration<'a>,
    import_name: &str,
) -> VariableDeclaration<'a> {
    // __oxjest_import_{}__ = await import("...")
    let await_import = ast.variable_declarator(
        Span::default(),
        VariableDeclarationKind::Const,
        ast.binding_pattern(
            ast.binding_pattern_kind_binding_identifier(Span::default(), import_name),
            Option::<Box<'a, _>>::None,
            false,
        ),
        Some(ast.expression_await(
            Span::default(),
            ast.expression_import(
                Span::default(),
                ast.expression_string_literal(decl.source.span, decl.source.value, decl.source.raw),
                ast.vec(),
                None,
            ),
        )),
        false,
    );

    // foo = __oxjest_import_{}__.foo, bar = __oxjest_import_{}__.default, ...
    let declarations = decl.specifiers.iter().flatten().map(|specifier| {
        ast.variable_declarator(
            Span::default(),
            VariableDeclarationKind::Const,
            ast.binding_pattern(
                BindingPatternKind::BindingIdentifier(
                    ast.alloc(specifier.local().clone_in(ast.allocator)),
                ),
                Option::<Box<'a, _>>::None,
                false,
            ),
            Some(match specifier {
                ImportDeclarationSpecifier::ImportNamespaceSpecifier(_) => {
                    ast.expression_identifier(Span::default(), import_name)
                }
                _ => ast
                    .member_expression_static(
                        Span::default(),
                        ast.expression_identifier(Span::default(), import_name),
                        match specifier {
                            ImportDeclarationSpecifier::ImportDefaultSpecifier(specifier) => {
                                ast.identifier_name(specifier.span, "default")
                            }
                            ImportDeclarationSpecifier::ImportSpecifier(specifier) => ast
                                .identifier_name(
                                    specifier.imported.span(),
                                    specifier.imported.name(),
                                ),
                            _ => unreachable!(),
                        },
                        false,
                    )
                    .into(),
            }),
            false,
        )
    });

    // const __oxjest_import_{}__ = await import("..."), foo = __oxjest_import_{}__.foo, ...;
    ast.variable_declaration(
        decl.span,
        VariableDeclarationKind::Const,
        ast.vec_from_iter(once(await_import).chain(declarations)),
        false,
    )
}

struct DynamicImportsVisitor<'a> {
    ast: AstBuilder<'a>,
    is_mock_found: AtomicBool,
    import_id: AtomicUsize,
}

impl<'a> VisitMut<'a> for DynamicImportsVisitor<'a> {
    fn visit_call_expression(&mut self, it: &mut CallExpression<'a>) {
        walk_call_expression(self, it);

        if is_jest_mock_call(it) || is_jest_mock_module_call(it) {
            self.is_mock_found.store(true, Ordering::Release);
        }
    }

    fn visit_statement(&mut self, it: &mut Statement<'a>) {
        walk_statement(self, it);

        let Statement::ImportDeclaration(decl) = &it else {
            return;
        };

        // Imports before any mocking don't need to be turned into dynamic imports
        if !self.is_mock_found.load(Ordering::Acquire) {
            return;
        }

        let import_id = self.import_id.fetch_add(1, Ordering::Relaxed);
        let import_name = format!("__oxjest_import_{}__", import_id);

        *it = Statement::VariableDeclaration(self.ast.alloc(make_dynamic_import(
            self.ast,
            decl,
            &import_name,
        )));
    }
}

/// Convert import declarations after module mocking to dynamic imports.
pub(crate) struct DynamicImports;

impl Pass for DynamicImports {
    fn process<'a>(&mut self, program: &mut Program<'a>, ast: AstBuilder<'a>) {
        DynamicImportsVisitor {
            ast,
            is_mock_found: AtomicBool::new(false),
            import_id: AtomicUsize::new(0),
        }
        .visit_program(program)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::transform;

    #[test]
    fn test_dynamic_imports() {
        let source_text = r#"
        // this import will kept as-is
        import { greeter } from "./greeter.js";

        jest.unstable_mockModule("./greeter.js", () => ({
          greet: () => "Hello, world!",
        }));

        // this import needs to be turned into dynamic one to be evaluated after the mocking above
        import { greet } from "./greeter.js";
        "#;

        let code = transform(source_text, DynamicImports);

        insta::assert_snapshot!(code, @r#"
        import { greeter } from "./greeter.js";
        jest.unstable_mockModule("./greeter.js", () => ({ greet: () => "Hello, world!" }));
        const __oxjest_import_0__ = await import("./greeter.js"), greet = __oxjest_import_0__.greet;
        "#);
    }
}
