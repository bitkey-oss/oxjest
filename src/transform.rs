use std::path::PathBuf;
use std::sync::Arc;

use base64::Engine;
use base64::prelude::BASE64_STANDARD;
use napi::bindgen_prelude::*;
use oxc::allocator::Allocator;
use oxc::codegen::{CodeGenerator, CodegenOptions, CodegenReturn};
use oxc_sourcemap::SourceMap;
use oxc_traverse::traverse_mut;

use crate::loader::Loader;
use crate::pass::Transformer;

pub(crate) fn _transform(
    source_text: String,
    source_path: String,
) -> Result<crate::TransformedSource> {
    let source_path = PathBuf::from(source_path);
    let allocator = Allocator::new();
    let (mut program, symbols, scopes) = Loader
        .load_str(&allocator, &source_text, &source_path)
        .map_err(|_| Error::from_reason("Could not load a source file. Invalid syntax?"))?;

    let mut transformer = Transformer::new();

    traverse_mut(&mut transformer, &allocator, &mut program, symbols, scopes);

    let CodegenReturn { mut code, map, .. } = CodeGenerator::new()
        .with_options(CodegenOptions {
            source_map_path: Some(source_path.clone()),
            ..Default::default()
        })
        .build(&program);

    // SAFETY: map is always Some(_) if CodegenOptions.source_map_path is Some(_)
    let map = map.unwrap();

    // Remove sourcesContent from the sourcemap as larger sources can lead OOM on Node.js
    let map = SourceMap::new(
        map.get_file().map(Arc::from),
        map.get_names().map(Arc::from).collect(),
        map.get_source_root().map(String::from),
        map.get_sources().map(Arc::from).collect(),
        None,
        map.get_tokens().cloned().collect(),
        None,
    );

    let map = map.to_json_string();

    // Append the source map to the code for better compatibility
    // https://github.com/swc-project/swc/blob/b22d7ee3ab8ee0a6dd521298237c42633137c633/crates/swc_compiler_base/src/lib.rs#L264
    code.push_str("\n//# sourceMappingURL=data:application/json;base64,");
    BASE64_STANDARD.encode_string(map.as_bytes(), &mut code);

    Ok(crate::TransformedSource { code, map })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    fn test(source_path: &Path) {
        let source_text = std::fs::read_to_string(source_path).unwrap();
        let source_path = source_path
            .file_name()
            .unwrap()
            .to_string_lossy()
            .to_string();

        let crate::TransformedSource { code, .. } =
            _transform(source_text, source_path.clone()).unwrap();

        insta::assert_snapshot!(source_path.as_str(), code);
    }

    test_each_file::test_each_path! { in "./tests" => test }
}
