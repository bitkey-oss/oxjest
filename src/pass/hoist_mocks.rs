use oxc::allocator::{Box, CloneIn, Vec};
use oxc::ast::AstBuilder;
use oxc::ast::ast::{Argument, Expression, Program, Span, Statement};
use oxc::ast_visit::VisitMut;
use oxc::ast_visit::walk_mut::{walk_expression, walk_program, walk_statements};

use crate::jest::is_jest_mock_call;
use crate::pass::Pass;

fn make_create_mock_factory<'a>(ast: AstBuilder<'a>, id: &str) -> Expression<'a> {
    ast.expression_call(
        Span::default(),
        ast.member_expression_static(
            Span::default(),
            ast.expression_identifier(Span::default(), "__oxjest__"),
            ast.identifier_name(Span::default(), "createMockFactory"),
            false,
        )
        .into(),
        Option::<Box<'_, _>>::None,
        ast.vec1(
            ast.expression_await(
                Span::default(),
                ast.expression_import(
                    Span::default(),
                    ast.expression_string_literal(Span::default(), id, None),
                    ast.vec(),
                    None,
                ),
            )
            .into(),
        ),
        false,
    )
}

struct HoistMocksVisitor<'a> {
    ast: AstBuilder<'a>,
    mocks: Vec<'a, Expression<'a>>,
}

impl<'a> VisitMut<'a> for HoistMocksVisitor<'a> {
    fn visit_program(&mut self, it: &mut Program<'a>) {
        walk_program(self, it);

        it.body.splice(
            0..0,
            self.mocks.iter().map(|mock_call| {
                self.ast
                    .statement_expression(Span::default(), mock_call.clone_in(self.ast.allocator))
            }),
        );
    }

    fn visit_expression(&mut self, it: &mut Expression<'a>) {
        walk_expression(self, it);

        let Expression::CallExpression(call) = it else {
            return;
        };

        if !is_jest_mock_call(call) {
            return;
        }

        let Expression::StaticMemberExpression(member) = &mut call.callee else {
            // SAFETY: Already checked above
            unreachable!();
        };

        member.property.name = self.ast.atom("unstable_mockModule");

        if call.arguments.len() < 2 {
            let Some(Argument::StringLiteral(lit)) = call.arguments.first() else {
                return;
            };

            let id = lit.value.as_str();

            call.arguments
                .push(make_create_mock_factory(self.ast, id).into())
        }

        self.mocks.push(self.ast.move_expression(it));
    }

    fn visit_statements(&mut self, it: &mut Vec<'a, Statement<'a>>) {
        walk_statements(self, it);

        it.retain(|stmt| !matches!(stmt, Statement::ExpressionStatement(stmt) if stmt.expression.is_null()));
    }
}

pub(crate) struct HoistMocks;

impl Pass for HoistMocks {
    fn process<'a>(&mut self, program: &mut Program<'a>, ast: AstBuilder<'a>) {
        HoistMocksVisitor {
            ast,
            mocks: Vec::new_in(ast.allocator),
        }
        .visit_program(program)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::transform;

    #[test]
    fn test_hoist_mocks() {
        let source_text = r#"
        import { greet } from "./greeter.js";

        // this mocking needs to be hoisted to the top of this module
        jest.mock("./greeter.js", () => ({
          greet: () => "Hello, world!",
        }));
        "#;

        let code = transform(source_text, HoistMocks);

        insta::assert_snapshot!(code, @r#"
        jest.unstable_mockModule("./greeter.js", () => ({ greet: () => "Hello, world!" }));
        import { greet } from "./greeter.js";
        "#);
    }
}
