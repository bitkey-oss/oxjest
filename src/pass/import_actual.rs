use oxc::allocator::Box;
use oxc::ast::AstBuilder;
use oxc::ast::ast::{Argument, Expression, Program, VariableDeclarationKind};
use oxc::span::{Atom, Span};
use oxc_traverse::{Traverse, TraverseCtx};

use crate::jest::is_jest_require_actual_call;

fn make_import_name(ast: AstBuilder, index: usize) -> &str {
    ast.str(&format!("__oxjest_actual_{index}__"))
}

/// Turn `jest.requireActual()` calls into dynamic imports, then hoists to the top of the module.
pub(crate) struct ImportActual<'a> {
    modules: Vec<Atom<'a>>,
}

impl ImportActual<'_> {
    pub(crate) fn new() -> Self {
        Self {
            modules: Vec::new(),
        }
    }
}

impl<'a> Traverse<'a> for ImportActual<'a> {
    fn exit_program(&mut self, node: &mut Program<'a>, ctx: &mut TraverseCtx<'a>) {
        if self.modules.is_empty() {
            return;
        }

        // Create `const __oxjest_actual_{}__ = await import(...), ...;` declaration
        let decl = ctx.ast.declaration_variable(
            Span::default(),
            VariableDeclarationKind::Const,
            ctx.ast
                .vec_from_iter(self.modules.iter().enumerate().map(|(index, id)| {
                    let await_import = ctx.ast.expression_await(
                        Span::default(),
                        ctx.ast.expression_import(
                            Span::default(),
                            ctx.ast.expression_string_literal(Span::default(), id, None),
                            ctx.ast.vec(),
                            None,
                        ),
                    );

                    ctx.ast.variable_declarator(
                        Span::default(),
                        VariableDeclarationKind::Const,
                        ctx.ast.binding_pattern(
                            ctx.ast.binding_pattern_kind_binding_identifier(
                                Span::default(),
                                make_import_name(ctx.ast, index),
                            ),
                            Option::<Box<'a, _>>::None,
                            false,
                        ),
                        Some(await_import),
                        false,
                    )
                })),
            false,
        );

        // Inject the declaration at the top of the module
        node.body.insert(0, decl.into());
    }

    fn exit_expression(&mut self, node: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
        let Expression::CallExpression(call) = node else {
            return;
        };

        if !is_jest_require_actual_call(call) {
            return;
        }

        // jest.requireActual("<id>");
        let Some(Argument::StringLiteral(lit)) = call.arguments.first() else {
            return;
        };

        let index = self.modules.len();
        self.modules.push(lit.value);

        *node = ctx
            .ast
            .expression_identifier(call.span, make_import_name(ctx.ast, index))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::transform;
    use oxc::allocator::Allocator;

    #[test]
    fn test_import_actual() {
        let source_text = r#"
        jest.unstable_mockModule("./greeter.js", () => ({
            ...jest.requireActual("./greeter.js"),
            greet: () => "Hello, world!",
        }));
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, ImportActual::new());

        insta::assert_snapshot!(code, @r#"
        const __oxjest_actual_0__ = await import("./greeter.js");
        jest.unstable_mockModule("./greeter.js", () => ({
        	...__oxjest_actual_0__,
        	greet: () => "Hello, world!"
        }));
        "#);
    }
}
