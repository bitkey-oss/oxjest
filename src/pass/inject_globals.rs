use oxc::allocator::Box;
use oxc::ast::AstBuilder;
use oxc::ast::ast::{
    Expression, IdentifierReference, ImportOrExportKind, MemberExpression, Program, Statement,
    StaticMemberExpression,
};
use oxc::span::Span;
use oxc_traverse::{Traverse, TraverseCtx};

const OXJEST_RUNTIME_ID: &str = "oxjest/runtime";
const OXJEST_RUNTIME_NAME: &str = "__oxjest__";
const JEST_OBJECT_NAME: &str = "jest";

fn make_runtime_import_stmt<'a>(ast: AstBuilder<'a>) -> Statement<'a> {
    Statement::ImportDeclaration(ast.alloc_import_declaration::<Option<Box<'a, _>>>(
        Span::default(),
        Some(
            ast.vec1(ast.import_declaration_specifier_import_namespace_specifier(
                Span::default(),
                ast.binding_identifier(Span::default(), OXJEST_RUNTIME_NAME),
            )),
        ),
        ast.string_literal(Span::default(), OXJEST_RUNTIME_ID, None),
        None,
        None,
        ImportOrExportKind::Value,
    ))
}

fn make_import_meta_jest<'a>(
    ast: AstBuilder<'a>,
    reference: &IdentifierReference<'a>,
) -> StaticMemberExpression<'a> {
    ast.static_member_expression(
        Span::default(),
        MemberExpression::StaticMemberExpression(ast.alloc_static_member_expression(
            Span::default(),
            ast.expression_identifier(Span::default(), "import"),
            ast.identifier_name(Span::default(), "meta"),
            false,
        ))
        .into(),
        ast.identifier_name(reference.span, reference.name),
        false,
    )
}

pub(crate) struct InjectGlobals {}

impl InjectGlobals {
    pub(crate) fn new() -> Self {
        Self {}
    }
}

impl<'a, State> Traverse<'a, State> for InjectGlobals {
    fn exit_program(&mut self, node: &mut Program<'a>, ctx: &mut TraverseCtx<'a, State>) {
        node.body.insert(0, make_runtime_import_stmt(ctx.ast));
    }

    fn enter_member_expression(
        &mut self,
        node: &mut MemberExpression<'a>,
        ctx: &mut TraverseCtx<'a, State>,
    ) {
        let MemberExpression::StaticMemberExpression(expr) = node else {
            return;
        };

        let Expression::Identifier(ident) = &expr.object else {
            return;
        };

        if ident.name != JEST_OBJECT_NAME {
            return;
        }

        expr.object = MemberExpression::StaticMemberExpression(
            ctx.ast.alloc(make_import_meta_jest(ctx.ast, ident)),
        )
        .into();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::transform;
    use oxc::allocator::Allocator;

    #[test]
    fn test_inject_globals() {
        let source_text = r#"
        // needs to be import.meta.jest
        jest.mock("./greeter.js");
        "#;

        let allocator = Allocator::new();
        let code = transform(&allocator, source_text, InjectGlobals::new());

        insta::assert_snapshot!(code, @r#"
        import * as __oxjest__ from "oxjest/runtime";
        // needs to be import.meta.jest
        import.meta.jest.mock("./greeter.js");
        "#);
    }
}
