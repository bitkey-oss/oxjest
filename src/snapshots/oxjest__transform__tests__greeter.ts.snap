---
source: src/transform.rs
expression: code
---
import * as __oxjest__ from "oxjest/runtime";
export function greet() {
	return "Hello, world!";
}
export class Greeter {
	static greetStatic() {
		return "Hello from static method!";
	}
	greet() {
		return "Hello from class method!";
	}
}

//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJuYW1lcyI6W10sInNvdXJjZXMiOlsiZ3JlZXRlci50cyJdLCJzb3VyY2VzQ29udGVudCI6W10sIm1hcHBpbmdzIjoiO0FBQUEsT0FBTyxTQUFTLFFBQWdCO0FBQzlCLFFBQU87QUFDUjtBQUVELE9BQU8sTUFBTSxRQUFRO0NBQ25CLE9BQU8sY0FBc0I7QUFDM0IsU0FBTztDQUNSO0NBRUQsUUFBZ0I7QUFDZCxTQUFPO0NBQ1I7QUFDRiJ9
